src/rust/src/zip339_ffi.rs

```diff
@@ -0,0 +1,110 @@
+use libc::{c_char, size_t};
+use std::{
+    convert::{TryFrom, TryInto},
+    ffi::{CStr, CString},
+    ptr, slice,
+};
+
+use zcash_primitives::zip339;
+
+// It's safer to use a wrapper type here than an enum. We can't stop a C caller from passing
+// an unrecognized value as a `language` parameter; if it were an enum on the Rust side,
+// then that would immediately be UB, whereas this way we can return null or false.
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct Language(pub u32);
+
+impl TryFrom<Language> for zip339::Language {
+    type Error = ();
+
+    fn try_from(language: Language) -> Result<Self, ()> {
+        // These must match `src/rust/include/zip339.h`.
+        match language {
+            Language(0) => Ok(zip339::Language::English),
+            Language(1) => Ok(zip339::Language::SimplifiedChinese),
+            Language(2) => Ok(zip339::Language::TraditionalChinese),
+            Language(3) => Ok(zip339::Language::Czech),
+            Language(4) => Ok(zip339::Language::French),
+            Language(5) => Ok(zip339::Language::Italian),
+            Language(6) => Ok(zip339::Language::Japanese),
+            Language(7) => Ok(zip339::Language::Korean),
+            Language(8) => Ok(zip339::Language::Portuguese),
+            Language(9) => Ok(zip339::Language::Spanish),
+            Language(_) => Err(()),
+        }
+    }
+}
+
+/// Creates a phrase with the given entropy, in the given `Language`. The phrase is represented as
+/// a pointer to a null-terminated C string that must not be written to, and must be freed by
+/// `zip339_free_phrase`.
+#[no_mangle]
+pub extern "C" fn zip339_entropy_to_phrase(
+    language: Language,
+    entropy: *const u8,
+    entropy_len: size_t,
+) -> *const c_char {
+    assert!(!entropy.is_null());
+
+    if let Ok(language) = language.try_into() {
+        let entropy = unsafe { slice::from_raw_parts(entropy, entropy_len) }.to_vec();
+        if let Ok(mnemonic) = zip339::Mnemonic::from_entropy_in(language, entropy) {
+            if let Ok(phrase) = CString::new(mnemonic.phrase()) {
+                return phrase.into_raw();
+            }
+        }
+    }
+    ptr::null()
```

Comment:
`phrase.into_raw()` returns `*mut c_char`:
```suggestion
    ptr::null_mut()
```